### PR TITLE
Wiki History Widget

### PR DESCRIPTION
**Pull request by heidiso**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidget.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	
+	@Override
+	public void configure(final WikiPageKey key, final boolean canEdit, final ActionHandler actionHandler) {
+		this.wikiKey = key;
+		synapseClient.getV2WikiHistory(wikiKey, new Long(100), new Long(0), new AsyncCallback<String>() {
+			@Override
+			public void onFailure(Throwable caught) {
+				if(!DisplayUtils.handleServiceException(caught, globalApplicationState.getPlaceChanger(), authenticationController.isLoggedIn(), view))
+					view.showErrorMessage(DisplayConstants.ERROR_LOADING_WIKI_HISTORY_WIDGET_FAILED+caught.getMessage());
```
Comment (dburdick) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidget.java:73]: instead write failed message into place where the wiki history would have been.

File: src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidget.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+					List<V2WikiHistorySnapshot> historyAsListOfHeaders = new ArrayList<V2WikiHistorySnapshot>();
+					for(int i = 0; i < historyAsList.size(); i++) {
+						V2WikiHistorySnapshot snapshot = (V2WikiHistorySnapshot) historyAsList.get(i);
+						historyAsListOfHeaders.add(snapshot);
+					}
+					// Update history data structure
+					view.updateHistoryList(historyAsListOfHeaders);
+					
+					// Prepare all user ids
+					List<String> userIds = new ArrayList<String>();
```
Comment (heidiso) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidget.java]: add user Ids to a set if not in the cache

File: src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidget.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+						@Override
+						public void onFailure(Throwable caught) {
+							view.showErrorMessage(DisplayConstants.ERROR_LOADING_WIKI_HISTORY_WIDGET_FAILED+caught.getMessage());
+						}
+						@Override
+						public void onSuccess(EntityWrapper result) {
+							try {
+								UserGroupHeaderResponsePage response = nodeModelCreator.createJSONEntity(result.getEntityJson(), UserGroupHeaderResponsePage.class);
+								// Store all the user display names
+								List<String> displayNames = new ArrayList<String>();
```
Comment (heidiso) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidget.java]: Add user headers as a local cache

File: src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidgetViewImpl.java [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+		loadMoreHistoryButton.addClickHandler(new ClickHandler() {
+			@Override
+			public void onClick(ClickEvent event) {
+		        offset += resultSize;
+		        presenter.configureNextPage(new Long(offset), new Long(10));
+			}
+			
+		});
+		// Disable if we've reached the end
+		if(hasReachedEndOfHistory) {
```
Comment (heidiso) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidgetViewImpl.java:173]: setVisible(false) to make it disappear

File: src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidgetViewImpl.java [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+			isFirstGetHistory = false;
+			createHistoryWidget();
+		} else {
+			historyTable.setRowCount(historyEntries.size());
+			// Set the range to display all of retrieved history so far
+		    historyTable.setVisibleRange(0, historyEntries.size());
+		    // Push in data
+		    historyTable.setRowData(0, historyEntries);
+		    // Disable button if no more history exists
+			if(hasReachedEndOfHistory) {
```
Comment (heidiso) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidgetViewImpl.java]: setVisible(false)

File: src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidgetViewImpl.java [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+		historyTable.addColumn(modifiedOnColumn, "Modified On");
+		
+		historyTable.setVisible(true);
+		historyTable.setRowCount(historyEntries.size());
+		
+		// Set the range to display in the table at one time
+	    historyTable.setVisibleRange(0, historyEntries.size());
+	    historyTable.setRowData(0, historyEntries);
+	    
+		FlowPanel panel = new FlowPanel();
```
Comment (heidiso) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidgetViewImpl.java]: setVisible(false) to hide flowPanel

File: src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidgetViewImpl.java [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+			public void invoke() {
+				actionHandler.restoreClicked(wikiVersion);
+			}	
+		};
+		org.sagebionetworks.web.client.utils.Callback cancelCallback = new org.sagebionetworks.web.client.utils.Callback() {
+			@Override
+			public void invoke() {
+			}	
+		};
+		DisplayUtils.showOkCancelMessage("Warning", "Are you sure you want to replace the current version with this one?", MessageBox.WARNING, 500, okCallback, cancelCallback);
```
Comment (heidiso) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/WikiHistoryWidgetViewImpl.java]: Add to DisplayConstants

File: src/main/webapp/Portal.css [hunk 8]
```diff
@@ -863,6 +863,24 @@ h1, h2, h3, h4, h5, h6 {
 	margin-bottom: 4px;
 }
 
+/** WikiPage Markdown History Widget **/
+.wikiHistoryWidget {
```
Comment (heidiso) [inline src/main/webapp/Portal.css]: remove because we're not centering anymore

**Pull request merged**